blocks/main_loop.py

```diff
@@ -42,6 +45,90 @@
 a model. Check documentation of your extensions."""
 
 
+class Timer(object):
+    def __init__(self):
+        self.total = defaultdict(int)
+        self.current = []
+        self.order = OrderedSet()
+
+    def enter(self, name):
+        self.current.append(name)
+        # We record the order in which sections were first called
+        self.order.add(tuple(self.current))
```

Comment:
I am not sure we need `OrderedSet` in Blocks only to support this one case. You can make `self.order` an `OrderedDict` and use `self.order[tuple(self.current)] = True` instead.